### PR TITLE
Energon ckpt multimodal

### PR DESCRIPTION
**Pull request by abhinavg4**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog

- Add specific line by line info of high level changes in this PR.

# Usage

- You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

File: nemo/collections/multimodal/data/energon/base.py [hunk 1]
```diff
@@ -101,6 +101,17 @@ def __init__(
         **kwargs: Additional keyword arguments. Will be passed to get_train_dataset() of Energon
         """
 
+        if not parallel_state.is_initialized():
+            logging.info(
+                f"Muiltimodal data loader parallel state is not initialized,"
+                f"Please ensure no data loader is being created on pipeline parallel ranks >0"
+            )
+        else:
+            pp_rank = parallel_state.get_pipeline_model_parallel_rank()
```
Comment (yaoyu-33) [inline nemo/collections/multimodal/data/energon/base.py:110]: what about other parallel like TP and EP?

File: requirements/requirements_multimodal.txt [hunk 2]
```diff
@@ -4,9 +4,9 @@ clip
 decord; sys_platform == 'linux'
 diffusers>=0.19.3
 einops_exts
+git+https://github.com/NVIDIA/Megatron-Energon.git
```
Comment (yaoyu-33) [inline requirements/requirements_multimodal.txt:7]: don't think use main is a good idea, at least pin to one of commit?

Comment (abhinavg4) [reply requirements/requirements_multimodal.txt:7]: So I think energon folks will release a new version soonish (Within next 24 hrs) and then I will update this to that version. I agree we should not use main.

File: nemo/collections/multimodal/data/energon/base.py [hunk 3]
```diff
@@ -294,10 +295,18 @@ def state_dict(self) -> Dict[str, Any]:
 
         if self.trainer:
             dataloader_obj = self.trainer.train_dataloader
-            state = dataloader_obj.save_state_global(dst_rank=0)
+
+            state = []
+            if torch.distributed.get_rank() == parallel_state.get_model_parallel_src_rank():
```
Comment (yashaswikarnati) [inline nemo/collections/multimodal/data/energon/base.py:300]: I would write a comment explaining the logic here for future purposes. Why we call get_model_parallel_src_rank and what energon is expected to return

Comment (yashaswikarnati) [inline nemo/collections/multimodal/data/energon/base.py:300]: if lets say tp =2, dp =2, pp =2. Lets say 0, 1, 2, 3 global ranks has MP group 0 and DP group 0. 4, 5, 6, 7 has MP group 1 and DP group . I'm assuming above condition with  get_model_parallel_src_rank() is True for rank 0 and 4. Is my understanding accurate? As they are different DP ranks I assume them to have different state but what is save_state_global doing? is it not doing gather from DP ranks?

Comment (abhinavg4) [reply nemo/collections/multimodal/data/energon/base.py:300]: I agree, I will add that

Comment (abhinavg4) [reply nemo/collections/multimodal/data/energon/base.py:300]: Yes you are right. Your understanding is correct. save_state_global saves the states from all the ranks.

Comment (yashaswikarnati) [reply nemo/collections/multimodal/data/energon/base.py:300]: https://github.com/NVIDIA/Megatron-LM/blob/e1586c218827fd686a7f88cb389fa0aff2fdb57e/megatron/core/dist_checkpointing/strategies/common.py#L40

Comment (yashaswikarnati) [reply nemo/collections/multimodal/data/energon/base.py:300]: nit; if rank 0 and rank 4 in the above example has same state, based on how megatron is saving common.pt it would be enough to return state from rank 0 I think.

Comment (abhinavg4) [reply nemo/collections/multimodal/data/energon/base.py:300]: Yes that's exacty what is ahppening but we need to make this call on every rank that has a unique state. 

`save_state_global` will return all the states to rank 0 and return None to all other ranks

Comment (yashaswikarnati) [reply nemo/collections/multimodal/data/energon/base.py:300]: Ah ok. thanks for the clarification. Thats a bit counter intuitive to call on multiple ranks but only one rank returns the full state. Please add a comment with this information.
